### PR TITLE
Add logging to ArchivePurger service, task + Scheduler class & modify CronArchive so scheduled Task logs are displayed.

### PR DESCRIPTION
**Pull request by diosmosis**
This pull request modifies CronArchive so scheduled task logging output displays in core:archive output, and adds logging to ArchivePurger + other classes.

We display scheduled task output in the core:archive process by going through the Request class instead of climulti:request.

Example scheduled task output:

```
INFO CoreConsole[2015-04-01 01:33:19] ---------------------------
INFO CoreConsole[2015-04-01 01:33:19] SCHEDULED TASKS
INFO CoreConsole[2015-04-01 01:33:19] Starting Scheduled tasks... 
INFO CoreAdminHome[2015-04-01 01:33:19] Scheduler: executing task Piwik\Plugins\CoreAdminHome\Tasks.purgeOutdatedArchives...
INFO CoreAdminHome[2015-04-01 01:33:19] Purging archives in 96 archive tables.
INFO CoreAdminHome[2015-04-01 01:33:19] No outdated archives found in archive numeric table for 2005-01-15.
INFO CoreAdminHome[2015-04-01 01:33:19] Purged 0 range archive rows from archive_numeric_2005_01 & archive_blob_2005_01.
INFO CoreAdminHome[2015-04-01 01:33:19] No outdated archives found in archive numeric table for 2006-01-15.
INFO CoreAdminHome[2015-04-01 01:33:19] Purged 0 range archive rows from archive_numeric_2006_01 & archive_blob_2006_01.
INFO CoreAdminHome[2015-04-01 01:33:19] No outdated archives found in archive numeric table for 2007-01-15.
INFO CoreAdminHome[2015-04-01 01:33:19] Scheduler: finished executing task Piwik\Plugins\CoreAdminHome\Tasks.purgeOutdatedArchives. Time elapsed: 0.114s
INFO CoreConsole[2015-04-01 01:33:19] done
INFO CoreConsole[2015-04-01 01:33:19] ---------------------------
```

Refs #7181

### CODE REVIEW

**Comment by diosmosis** [comment]
Remembered the CronArchive method is used within tracker, need to revise how tracker runs scheduled tasks.

**Comment by mattab** [comment]
> Remembered the CronArchive method is used within tracker

I'm wondering, do we have a test case for this? if not, it would be useful to add one   :+1:

File: core/Archive/ArchivePurger.php [hunk 1]
```diff
@@ -166,9 +198,11 @@ protected function deleteArchiveIds(Date $date, $idArchivesToDelete)
         $numericTable = ArchiveTableCreator::getNumericTable($date);
         $blobTable    = ArchiveTableCreator::getBlobTable($date);
 
+        $deletedCount = 0;
         foreach ($batches as $idsToDelete) {
-            $this->model->deleteArchiveIds($numericTable, $blobTable, $idsToDelete);
+            $deletedCount += $this->model->deleteArchiveIds($numericTable, $blobTable, $idsToDelete);
         }
+        return $deletedCount;
```
Comment (tsteur) [inline core/Archive/ArchivePurger.php]: Do you mind adding an integration test for this? Would be nice to have a test for this. Not only for the return value but also whether it actually deletes the archives.

File: core/CronArchive.php [hunk 2]
```diff
@@ -429,13 +429,12 @@ public function runScheduledTasks()
...
         $this->log("Starting Scheduled tasks... ");
 
-        $tasksOutput = $this->request("?module=API&method=CoreAdminHome.runScheduledTasks&format=csv&convertToUnicode=0&token_auth=" . $this->token_auth);
+        API\Request::processRequest("CoreAdminHome.runScheduledTasks", array(
+            'token_auth' => $this->token_auth,
+            'format' => 'original', // so exceptions get thrown
+            'trigger' => 'archivephp'
+        ));
 
-        if ($tasksOutput == \Piwik\DataTable\Renderer\Csv::NO_DATA_AVAILABLE) {
```
Comment (tsteur) [inline core/CronArchive.php]: Can you here check maybe for the response of the `runScheduledTasks` method and log this message if `$exectionResults` is an empty array? But I think I saw you added some logger statements in the Scheduler as well so maybe it is not needed here anymore :) Ignore it in this case :)

File: core/DataAccess/Model.php [hunk 3]
```diff
@@ -118,13 +129,21 @@ public function deleteArchivesWithPeriod($numericTable, $blobTable, $period, $da
...
+            $deletedRows += $queryObj->rowCount();
         } catch (Exception $e) {
             // Individual blob tables could be missing
+            $this->logger->debug("Unable to delete archives by period from {blobTable}.", array(
+                'blobTable' => $blobTable,
+                'exception' => $e
+            ));
         }
+
+        return $deletedRows;
```
Comment (tsteur) [inline core/DataAccess/Model.php]: Do you mind adding a simple test that it actually returns the number of deletedRows?

**Comment by tsteur** [comment]
Nice! Do those kinda changes affect performance in any way?

File: core/DataAccess/Model.php (hunk 3 above)
Comment (diosmosis) [inline core/DataAccess/Model.php]: @tsteur You mean an integration test that adds data to the DB and deletes it?

Comment (tsteur) [inline core/DataAccess/Model.php]: Yes. I wouldn't mock the DB connection. An integration test should be fine and you could check whether it actually returns the deleted count and make sure it only deletes the ones older than the date given

File: core/Archive/ArchivePurger.php (hunk 1 above)
Comment (diosmosis) [inline core/Archive/ArchivePurger.php]: Modified the existing ArchivePurgerTest to test for deleted row counts.

File: core/DataAccess/Model.php (hunk 3 above)
Comment (diosmosis) [inline core/DataAccess/Model.php]: FYI, I added tests to the existing ArchivePurgerTest. I think tests for just the Model would be somewhat redundant considering all it does is execute SQL.

Comment (tsteur) [inline core/DataAccess/Model.php]: > I think tests for just the Model would be somewhat redundant considering all it does is execute SQL.

It sounds like there could be no bugs, and like nobody would ever want to change that code :)

**Comment by diosmosis** [comment]
@tsteur Re: performance, I tested using a dump of demo2's DB (which admittedly wouldn't have too much to purge) and there was no discernible difference w/ logging added.

**Comment by diosmosis** [comment]
Ready for another review, and then possibly merge. Now includes system test to ensure scheduled tasks are run via tracker correctly and w/o failing the tracker.

**Comment by mnapoli** [comment]
I have checked out the branch and given it a try, the amount of output when running the scheduled tasks is quite high. For example:

```
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: finished executing task Piwik\Plugins\CoreAdminHome\Tasks.purgeOutdatedArchives. Time elapsed: 0.219s
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: executing task Piwik\Plugins\ExamplePlugin\Tasks.myTaskWithParam_anystring...
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: finished executing task Piwik\Plugins\ExamplePlugin\Tasks.myTaskWithParam_anystring. Time elapsed: 0.000s
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: executing task Piwik\Plugins\ExamplePlugin\Tasks.myTask...
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: finished executing task Piwik\Plugins\ExamplePlugin\Tasks.myTask. Time elapsed: 0.000s
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: executing task Piwik\Plugins\ExamplePlugin\Tasks.myTask...
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: finished executing task Piwik\Plugins\ExamplePlugin\Tasks.myTask. Time elapsed: 0.000s
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: executing task Piwik\Plugins\ExamplePlugin\Tasks.myTask...
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: finished executing task Piwik\Plugins\ExamplePlugin\Tasks.myTask. Time elapsed: 0.000s
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: executing task Piwik\Plugins\ExamplePlugin\Tasks.myTask...
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: finished executing task Piwik\Plugins\ExamplePlugin\Tasks.myTask. Time elapsed: 0.000s
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: executing task Piwik\Plugins\ExamplePlugin\Tasks.myTaskWithParam_anystring...
INFO CoreAdminHome[2015-04-16 03:56:52] Scheduler: finished executing task Piwik\Plugins\ExamplePlugin\Tasks.myTaskWithParam_anystring. Time elapsed: 0.000s
[...]
INFO CoreAdminHome[2015-04-16 03:58:37] done
Piwik\Plugins\CoreAdminHome\Tasks.purgeOutdatedArchives - Time elapsed: 0.219s
Piwik\Plugins\ExamplePlugin\Tasks.myTaskWithParam_anystring - Time elapsed: 0.000s
Piwik\Plugins\ExamplePlugin\Tasks.myTask - Time elapsed: 0.000s
Piwik\Plugins\ExamplePlugin\Tasks.myTask - Time elapsed: 0.000s
Piwik\Plugins\ExamplePlugin\Tasks.myTask - Time elapsed: 0.000s
Piwik\Plugins\ExamplePlugin\Tasks.myTask - Time elapsed: 0.000s
Piwik\Plugins\ExamplePlugin\Tasks.myTaskWithParam_anystring - Time elapsed: 0.000s
Piwik\Plugins\ScheduledReports\API.sendReport_1 - Time elapsed: 103.436s
Piwik\Plugins\CustomAlerts\Tasks.runAlertsDaily_16 - Time elapsed: 0.119s
Piwik\Plugins\CustomAlerts\Tasks.runAlertsWeekly_16 - Time elapsed: 0.002s
Piwik\Plugins\CustomAlerts\Tasks.runAlertsMonthly_16 - Time elapsed: 0.003s
Piwik\Plugins\TrafficMonitor\Tasks.doCacheTotalActionsNbFromRawData - Time elapsed: 0.962s
Piwik\Plugins\CoreAdminHome\Tasks.purgeInvalidatedArchives - Time elapsed: 0.002s
Piwik\Plugins\PrivacyManager\Tasks.deleteReportData - Time elapsed: 0.004s
Piwik\Plugins\PrivacyManager\Tasks.deleteLogData - Time elapsed: 0.001s
Piwik\Plugins\CorePluginsAdmin\Tasks.clearAllCacheEntries - Time elapsed: 0.001s
Piwik\Plugins\CorePluginsAdmin\Tasks.sendNotificationIfUpdatesAvailable - Time elapsed: 0.001s
Piwik\Plugins\CoreAdminHome\Tasks.optimizeArchiveTable - Time elapsed: 0.028s
Piwik\Plugins\ExamplePlugin\Tasks.myTask - Time elapsed: 0.000s
Piwik\Plugins\UserCountry\GeoIPAutoUpdater.update - Time elapsed: 0.011s
Piwik\Plugins\CoreUpdater\Tasks.sendNotificationIfUpdateAvailable - Time elapsed: 0.001s
```

Would it be better to remove the `finished executing ...` kind of messages? E.g. only say `executing task …` and not say when it's finished (because it will be obvious from the output).

Or else get rid of the summary at the end (which only displays the time elapsed) and show the time elapsed when printing `finished executing task ...`?

File: plugins/CoreAdminHome/Commands/PurgeOldArchiveData.php [hunk 4]
```diff
@@ -68,7 +69,12 @@ protected function configure()
 
     protected function execute(InputInterface $input, OutputInterface $output)
     {
-        $archivePurger = $this->archivePurger;
+        $logger = null;
+        if ($output->getVerbosity() <= OutputInterface::VERBOSITY_NORMAL) {
+            $logger = new NullLogger();
+        }
```
Comment (mnapoli) [inline plugins/CoreAdminHome/Commands/PurgeOldArchiveData.php]: Why this piece of code? The `-q` flag should automatically set the error level at "error".

Comment (diosmosis) [inline plugins/CoreAdminHome/Commands/PurgeOldArchiveData.php]: During normal output (in addition to when `-q` is used), we don't want the ArchivePurger logs to output in this command, but we want it to appear in scheduled task output.

**Comment by diosmosis** [comment]
> Or else get rid of the summary at the end (which only displays the time elapsed) and show the time elapsed when printing finished executing task ...?

You're right, don't need the summary anymore, will remove it.

File: plugins/CoreAdminHome/Commands/PurgeOldArchiveData.php (hunk 4 above)
Comment (mnapoli) [inline plugins/CoreAdminHome/Commands/PurgeOldArchiveData.php]: Ah yes I think we discussed that in slack! Could you maybe add a comment to explain it?

Comment (diosmosis) [inline plugins/CoreAdminHome/Commands/PurgeOldArchiveData.php]: Added a comment.

**Comment by mattab** [comment]
> only say executing task … and not say when it's finished (because it will be obvious from the output).

it looks like the output was not simplfied as much as possible so created follow up issue https://github.com/piwik/piwik/issues/7705

**Pull request merged**